zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs

```diff
@@ -111,13 +177,85 @@ impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+                        panic!("A Sapling output must exist at index {}", output_index)
+                    });
+
+                let pivk = PreparedIncomingViewingKey::new(&dfvk.to_ivk(Scope::Internal));
+                if try_sapling_note_decryption(&pivk, output, zip212_enforcement).is_some() {
+                    transaction.execute(
+                        "UPDATE sapling_received_notes SET recipient_key_scope = :scope
+                         WHERE id_note = :note_id",
+                        named_params! {":scope": scope_code(Scope::Internal), ":note_id": note_id},
+                    )?;
+                }
+            } else {
+                let diversifier = {
+                    let d: Vec<_> = row.get(6)?;
+                    if d.len() != 11 {
+                        return Err(WalletMigrationError::CorruptedData(
+                            "Invalid diversifier length".to_string(),
+                        ));
+                    }
+                    let mut tmp = [0; 11];
+                    tmp.copy_from_slice(&d);
+                    Diversifier(tmp)
+                };
+
+                let note_value =
+                    NonNegativeAmount::from_nonnegative_i64(row.get(7)?).map_err(|_e| {
+                        WalletMigrationError::CorruptedData(
+                            "Note values must be nonnegative".to_string(),
+                        )
+                    })?;
+
+                let rseed = {
+                    let rcm_bytes: [u8; 32] =
+                        row.get::<_, Vec<u8>>(8)?[..].try_into().map_err(|_| {
+                            WalletMigrationError::CorruptedData(format!(
+                                "Note {} is invalid",
+                                note_id
+                            ))
+                        })?;
+
+                    let rcm = Option::from(jubjub::Fr::from_repr(rcm_bytes)).ok_or(
+                        WalletMigrationError::CorruptedData(format!("Note {} is invalid", note_id)),
+                    )?;
+
+                    // The wallet database always stores the `rcm` value, and not `rseed`,
+                    // so for note reconstruction we always use `BeforeZip212`.
+                    Rseed::BeforeZip212(rcm)
+                };
+
+                let note_commitment_tree_position =
+                    Position::from(u64::try_from(row.get::<_, i64>(9)?).map_err(|_| {
+                        WalletMigrationError::CorruptedData(
+                            "Note commitment tree position invalid.".to_string(),
+                        )
+                    })?);
+
+                let scope = select_note_scope(
+                    &mut commitment_tree,
+                    dfvk,
+                    &diversifier,
+                    &note_value.try_into().unwrap(),
+                    &rseed,
+                    note_commitment_tree_position,
                 )?;
+
+                if scope == Some(Scope::Internal) {
```

Comment:
In the case where `scope == None`, it means that we have no way to spend the note, so the scope of the key is irrelevant. We _could_ add a sentinel for this, but I think that just leaving it as `External` is probably okay?